src/bootstrap/install.rs

```diff
@@ -175,10 +175,12 @@ install!((self, builder, _config),
         install_docs(builder, self.stage, self.target);
     };
     Std, "src/libstd", true, only_hosts: true, {
-        builder.ensure(dist::Std {
-            compiler: builder.compiler(self.stage, self.host),
-            target: self.target
-        });
+        for target in &builder.build.targets {
```

Comment:
This feels wrong to me. In general, we shouldn't have to think about targets within build steps, instead the step should be configured to run for all targets. Perhaps the actual change should be on the previous line, where `only_hosts: true` is currently -- maybe that should be `false`? I'm not too certain what we're trying to accomplish here, either, but mostly am happy to take changes as long as @alexcrichton is happy.

Reply:
That may be better, however I was not sure of the exact meaning of that flag. Let me try to explain this commit: I was only mirroring the already-existing logic in `install_std` further above, which explicitly calls `install_sh` over all targets. Because it does this, I thought that it should also explicitly call `ensure(dist::Std ...)` explicitly over all targets.

You are right that it would be neater to do this declaratively though, to declare "depends on all architectures" on a meta-level outside of the step definitions. This would also allow (and maybe require) `install_std` to not loop over all target arches.

If I understand correctly, the overall goal of "x.py install" is to install all targets' `libstd` into `DESTDIR`. Now there is a further issue that actually rustbuild supports multiple hosts, so in fact this process is run for all hosts, installing into the same `DESTDIR`, which might explain the 2x2 issue I described in my other post. In more detail:

From the documentation of config.toml.example, it seems that rust turns my config
~~~~
build = "x86_64-unknown-linux-gnu"
host = ["powerpc64le-unknown-linux-gnu"]
target = ["powerpc64le-unknown-linux-gnu"]
~~~~
effectively into
~~~~
build = "x86_64-unknown-linux-gnu"
host = [build + "powerpc64le-unknown-linux-gnu"]
target = host + ["powerpc64le-unknown-linux-gnu"]
~~~~

and this is what gets run in the "install" process. (Hopefully someone can confirm; I'm just guessing here based on what the documentation says.) So "install" gets run for all hosts, and each of these installs all targets' libstd, at least after my patch is applied to fix the errors.

Obviously this is not ideal (multiple `/usr/bin/rustc` gets installed into the same location, etc), but this currently does not affect me negatively, so I'm prepared to defer the solution to this for another day - since the $build architecture install process gets run first, the foreign architecture (what I'm actually interested in) ends up clobbering all the previous files, and I end up with what I actually want. I'm also not interested in the $build-architecture libstd, but that gets installed into a architecture-specific path so I can just ignore it.

Reply:
That all seems correct to me -- I'm happy to take this for now, I think, since it seems to be following the pre-existing logic we have in the install step anyway. Going to leave to @alexcrichton's final approval, but sounds good to me.